sympy/combinatorics/perm_groups.py

```diff
@@ -2288,6 +2288,14 @@ def minimal_block(self, points):
             # force path compression to get the final state of the equivalence
             # relation
             self._union_find_rep(i, parents)
+
+        # rewrite result so that block representatives are minimal
+        new_reps = {}
```

Comment:
How about:

```python
saw = {}  # record position at which k was first seen
return [saw.setdefault(r, i) for i, r in enumerate(parents)]
```

Reply:
That is a nice trick (though not easy to decipher :)

Reply:
@smichr That's quite cool, thanks! :)